File: src/main/java/com/hp/jipp/encoding/KeywordType.java
package com.hp.jipp.encoding;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import com.hp.jipp.util.Util;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.util.Collection;
import java.util.Map;

public class KeywordType<T extends Keyword> extends AttributeType<T> {

    /** An encoder for Keyword types */
    public static class Encoder<T extends Keyword> extends Attribute.Encoder<T> {

        @SuppressWarnings("unchecked")
        public static <T extends Keyword> Encoder<T> of(Class<T> cls, Keyword.Factory<T> factory) {
            ImmutableList.Builder<T> all = new ImmutableList.Builder<>();
            for (Object object : Util.getStaticObjects(cls)) {
                if (cls.isAssignableFrom(object.getClass())) {
                    //noinspection ResultOfMethodCallIgnored
                    all.add((T) object);
                }
            }
            return new Encoder(factory, all.build(), cls.getSimpleName());
        }

        private final Map<String, T> map;
        private final Keyword.Factory<T> factory;

        public Encoder(Keyword.Factory<T> factory, Collection<T> all, String name) {
            super(name);
            ImmutableMap.Builder<String, T> builder = new ImmutableMap.Builder<>();
            for (T keyword : all) {
                //noinspection ResultOfMethodCallIgnored
                builder.put(keyword.getName(), keyword);
            }
            map = builder.build();
            this.factory = factory;
        }

        @Override
        public T readValue(DataInputStream in, Tag valueTag) throws IOException {
            String value = StringType.ENCODER.readValue(in, valueTag);
            if (map.containsKey(value)) {
                return map.get(value);
            } else {
                return factory.of(value);
            }
        }

        @Override
        public void writeValue(DataOutputStream out, T value) throws IOException {
            StringType.ENCODER.writeValue(out, value.getName());
        }

        @Override
        public boolean valid(Tag valueTag) {
            return valueTag == Tag.Keyword;
        }

        public Collection<T> getAll() {
            return map.values();
        }
    }

    public KeywordType(Encoder<T> encoder, String name) {
        super(encoder, Tag.Keyword, name);
    }
}

File: src/main/java/com/hp/jipp/encoding/KeywordType.kt
// Copyright 2017 HP Development Company, L.P.
// SPDX-License-Identifier: MIT

package com.hp.jipp.encoding

import com.hp.jipp.util.getStaticObjects
import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.IOException

/** Attribute type for attributes based on [Keyword] */
open class KeywordType<T : Keyword>(encoder: KeywordType.Encoder<T>, override val name: String) :
        AttributeType<T>(encoder, Tag.keyword) {

    /** An encoder for keyword types  */
    class Encoder<T : Keyword>(private val factory: Keyword.Factory<T>, all: Collection<T>, name: String) :
            SimpleEncoder<T>(name) {

        private val map: Map<String, T> = all.map { it.name to it }.toMap()

        @Throws(IOException::class)
        override fun readValue(input: DataInputStream, valueTag: Tag): T {
            val key = StringType.ENCODER.readValue(input, valueTag)
            return map[key] ?: factory.of(key)
        }

        @Throws(IOException::class)
        override fun writeValue(out: DataOutputStream, value: T) {
            StringType.ENCODER.writeValue(out, value.name)
        }

        override fun valid(valueTag: Tag) = valueTag == Tag.keyword

        val all: Collection<T>
            get() = map.values
    }

    companion object {
        /** Return a new [Encoder] for a class internally defining static [Keyword] objects */
        fun <T : Keyword> encoderOf(cls: Class<T>, factory: Keyword.Factory<T>): KeywordType.Encoder<T> =
                KeywordType.Encoder(factory, cls.getStaticObjects()
                        .filter { cls.isAssignableFrom(it.javaClass) }
                        .map {
                            @Suppress("UNCHECKED_CAST")
                            it as T
                        }, cls.simpleName)

        /** Return a new [Encoder] for a class internally defining static [Keyword] objects */
        fun <T : Keyword> encoderOf(cls: Class<T>, factory: (String) -> T): KeywordType.Encoder<T> =
                encoderOf(cls, object : Keyword.Factory<T> {
                    override fun of(name: String): T = factory(name)
                })
    }
}
